Align torch.library.impl with the new torch.library style

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111310
* #111380
* #111309
* __->__ #111308
* #111307

We add a new overload to torch.library.impl that accepts an optional
Library arg. If provided, the lifetime of the registration will be
tied to the Library arg, otherwise, it will live forever.

Test Plan:
- existing and new tests

zou3519 committed:
Align torch.library.impl with the new torch.library style

We add a new overload to torch.library.impl that accepts an optional
Library arg. If provided, the lifetime of the registration will be
tied to the Library arg, otherwise, it will live forever.

Test Plan:
- existing and new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Align torch.library.impl with the new torch.library style"


We add a new overload to torch.library.impl that accepts an optional
Library arg. If provided, the lifetime of the registration will be
tied to the Library arg, otherwise, it will live forever.

Test Plan:
- existing and new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Align torch.library.impl with the new torch.library style"


We add a new overload to torch.library.impl that accepts an optional
Library arg. If provided, the lifetime of the registration will be
tied to the Library arg, otherwise, it will live forever.

Test Plan:
- existing and new tests

[ghstack-poisoned]

soulitzer commented on torch/library.py:
```diff
@@ -245,6 +235,63 @@ def wrap(f):
+            use_lib = Library(namespace, "FRAGMENT")
+            _keep_alive.append(use_lib)
+        else:
+            use_lib = lib
```
Do we want to add to `_registration_handles` like we do for `impl_abstract`?

zou3519 replied:
We need `_registration_handles` for impl_abstract because there is otherwise no way to delete an impl_abstract because it is not entirely managed by the PyTorch dispatcher.

All of the `impl` registrations are managed by the PyTorch dispatcher, so they do not need `_registration_handles`

soulitzer commented on torch/library.py:
```diff
@@ -245,6 +235,63 @@ def wrap(f):
+    """Register an implementation for a DispatchKey for this operator.
+
+    torch.library.impl is a low-level API to directly register
+    an implementation for an operator for some DispatchKey in PyTorch.
```
maybe direct people to the high level api for devices

zou3519 committed:
Update on "Align torch.library.impl with the new torch.library style"


We add a new overload to torch.library.impl that accepts an optional
Library arg. If provided, the lifetime of the registration will be
tied to the Library arg, otherwise, it will live forever.

Test Plan:
- existing and new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
